b2sdk/transfer/emerge/unbound_write_intent.py

```diff
@@ -0,0 +1,230 @@
+        ``queue_size`` is handled on a best-effort basis. It's possible, in rare cases, that there will be more buffers
+        available at once. With current implementation that would be the case when the whole buffer was read, but on
+        the very last byte the server stopped responding and a retry is issued.
+
+        :param read_only_source: Python object that has a ``read`` method.
+        :param buffer_size_bytes: Size of a single buffer that we're to download from the source and push to the cloud.
+        :param read_size: Size of a single read to be performed on ``read_only_source``.
+        :param queue_size: Maximal amount of buffers that will be created.
+        :param queue_timeout_seconds: Iterator will wait at most this many seconds for an empty slot
+                                      for a buffer. After that time it's considered an error.
+        """
+        assert queue_size >= 1 and read_size > 0 and buffer_size_bytes > 0 and queue_timeout_seconds > 0.0
+
+        self.read_only_source = read_only_source
+        self.read_size = read_size
+
+        self.buffer_size_bytes = buffer_size_bytes
+        self.buffer_limit_queue = queue.Queue(maxsize=queue_size)
+        self.queue_timeout_seconds = queue_timeout_seconds
+
+        self.buffer = bytearray()
+        self.leftovers_buffer = bytearray()
+
+    def iterator(self) -> Iterator[WriteIntent]:
+        """
+        Creates new ``WriteIntent`` objects as the data is pulled from the ``read_only_source``.
+        """
+        datastream_done = False
+        offset = 0
+
+        while not datastream_done:
+            self._wait_for_free_buffer_slot()
+
+            # In very small buffer sizes and large read sizes we could
+            # land with multiple buffers read at once. This should happen
+            # only in tests.
+            self._trim_to_leftovers()
+
+            while len(self.buffer) < self.buffer_size_bytes:
+                data = self.read_only_source.read(self.read_size)
+                if len(data) == 0:
+                    datastream_done = True
+                    break
+
+                self.buffer += data
+                self._trim_to_leftovers()
+
+            # If we've just started a new buffer and got an empty read on it,
+            # we have no data to send and the process is finished.
+            if len(self.buffer) == 0:
+                self._release_buffer()
+                break
+
+            source = UnboundSourceBytes(self.buffer, self._release_buffer)
+            intent = WriteIntent(source, destination_offset=offset)
+            yield intent
+
+            offset += len(self.buffer)
+            self._rotate_leftovers()
+
+    def _trim_to_leftovers(self) -> None:
+        if len(self.buffer) <= self.buffer_size_bytes:
+            return
+        remainder = len(self.buffer) - self.buffer_size_bytes
+        self.leftovers_buffer += self.buffer[-remainder:]
+        self.buffer = self.buffer[:-remainder]
```

Comment:
here you are doing memcopy, but I think instead you could use some kind of a view which is essentially a wrapper over an existing memory structure that pretends it has less data than it actually has. This would allow you to not copy blocks of data but instead read only the part that is interesting now

Reply:
Checked whether `memoryview` would grant any performance boost. It seems that it even creating and dumping the object speeds up this slicing operation by a factor of ~2.